airflow/api/client/stable_api_client.py

```diff
@@ -0,0 +1,114 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from json.decoder import JSONDecodeError
+from typing import Optional
+from urllib.parse import urljoin
+
+from airflow.api.client import api_client
+from airflow.models import DagRun
+from airflow.utils.state import State
+
+
+class AirflowClient(api_client.Client):
+    """API Client using the stable REST API"""
+
+    def _request(self, url, method='GET', json=None):
+        params = {
+            'url': url,
+        }
+        if json is not None:
+            params['json'] = json
+        resp = getattr(self._session, method.lower())(**params)
+        if resp.is_error:
+            # It is justified here because there might be many resp types.
+            try:
+                data = resp.json()
+            except Exception:
+                data = {}
+            raise OSError(data.get('error', 'Server error'))
+        try:  # DELETE requests doesn't return a value
+            return resp.json()
+        except JSONDecodeError:
+            return resp.text
```

Comment:
Instead of trying to handle different methods differently, I’d just have a helper function `parse_response` that returns `str` (and raises on exception), and let the caller methods call `self._session` themselves. You can actually just do `self._session(method, url, ...)` directly (methods like `self._session.get()` are actually just convenience wrappers).